sklearn/decomposition/tests/test_nmf.py

```diff
@@ -932,3 +932,50 @@ def test_NMF_inverse_transform_W_deprecation():
 
     with pytest.warns(FutureWarning, match="Input argument `W` was renamed to `Xt`"):
         est.inverse_transform(W=Xt)
+
+
+@pytest.mark.parametrize("Estimator", [NMF, MiniBatchNMF])
+def test_nmf_n_components_auto(Estimator):
+    rng = np.random.RandomState(0)
+    X = rng.random_sample((6, 5))
+    W = rng.random_sample((6, 2))
+    H = rng.random_sample((2, 5))
+    est = Estimator(
+        n_components="auto",
+        init="custom",
+        random_state=0,
+        tol=1e-6,
+    )
+    est.fit_transform(X, W=W, H=H)
+    assert est._n_components == 2
+
+
+def test_nmf_non_negative_factorization_n_components_auto():
+    rng = np.random.RandomState(0)
+    X = rng.random_sample((6, 5))
+    W_init = rng.random_sample((6, 2))
+    H_init = rng.random_sample((2, 5))
+    W, H, _ = non_negative_factorization(
+        X, W=W_init, H=H_init, init="custom", n_components="auto"
+    )
+    assert H.shape[0] == 2
+    assert W.shape[1] == 2
+
+
+def test_nmf_n_components_default_value_warning():
+    rng = np.random.RandomState(0)
+    X = rng.random_sample((6, 5))
+    H = rng.random_sample((2, 5))
+    with pytest.warns(
+        FutureWarning, match="The default value of `n_components` will change from"
+    ):
+        non_negative_factorization(X, H=H)
+
+
+def test_nmf_n_components_auto_no_h_update():
+    rng = np.random.RandomState(0)
+    H_true = rng.random_sample((2, 5))
+    X = rng.random_sample((6, 5))
+    _, H, _ = non_negative_factorization(
+        X, H=H_true, n_components="auto", update_H=False
+    )  # should not fail
```

Comment:
you can check that H.shape has not changed `H.shape == H_true.shape`.
you can also check that W has the appropriate shape.